zsh: add zproof option

MayNiklas opened the pull request:
### Description

<!--

Please provide a brief description of your change.

-->

Adds `programs.zsh.enableZproof`.
zproof is used to benchmark the shell initialisation -> helps to reduce loading time.
Needs to be the first & last line in `~/.zshrc`.
Since there isn't a `initExtraLast` option currently, it's hard to implement that without this option.

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

<!--
If you are updating a module, please @ people who are in its `meta.maintainers` list.
If in doubt, check `git blame` for whoever last touched something.
-->

I haven't found a `meta.maintainers` for zsh. Can somebody help me there?

teto commented:
> Since there isn't a initExtraLast option currently, it's hard to implement that without this option.

IIRC, I was able to do that with `lib.mkAfter` commands.

MayNiklas commented:
> > Since there isn't a initExtraLast option currently, it's hard to implement that without this option.
> 
> 
> 
> IIRC, I was able to do that with `lib.mkAfter` commands. 

I didn't knew about this!
Do you think my PS still makes sense? Or should I close it?

rycee commented on modules/programs/zsh.nix:
```diff
@@ -356,6 +356,13 @@ in
         description = "Enable zsh autosuggestions";
       };
 
+      enableZproof = mkOption {
```
I would suggest

```suggestion
      zproof.enable = mkOption {
```

rycee commented on modules/programs/zsh.nix:
```diff
@@ -534,6 +541,12 @@ in
       home.file."${relToDotDir ".zshrc"}".text = concatStringsSep "\n" ([
+        # zproof must be loaded before everything else, since it
+        # benchmarks the shell initialization.
+        (optionalString (cfg.enableZproof) ''
```
```suggestion
        (optionalString cfg.enableZproof ''
```

rycee commented on modules/programs/zsh.nix:
```diff
@@ -657,6 +670,11 @@ in
           }
         '')
+
+        (optionalString (cfg.enableZproof)
```
```suggestion
        (optionalString cfg.enableZproof
```

rycee commented:
I don't have any opinion on whether this is useful enough to add, just added a few minor comments 🙂

MayNiklas commented:
> I don't have any opinion on whether this is useful enough to add, just added a few minor comments 🙂

Thank you!
I implemented those.

In my personal opinion, having zproof easily accessible is helpful.
Especially with home manager, it's easy to have a bloated shell config, that takes long to initialise.
Having an easy way to debug initialisation time might be helpful in some scenarios.

rycee commented:
Sounds good. I reworded the commit message and rebased to master in 77c698fa4b3081b6019ad77d1bfedf06dbbde0db. Thanks for the contribution!

mrBliss commented:
Why did you call it zproof instead of zprof?

MayNiklas commented:
> Why did you call it zproof instead of zprof?

Human error 😬
@rycee What is easier? You fixing it or me opening another PR?

teto commented:
@rycee does too much for HM already, please open an MR if you can. I dont think we need an alias yet, just fix the spelling
